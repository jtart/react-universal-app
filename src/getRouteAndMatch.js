import { matchRoutes } from 'react-router-config';

const getRouteAndMatch = (url, routes) => {
  const matchedRoutes = matchRoutes(routes, url);
  return matchedRoutes.length > 0 ? matchedRoutes[0] : {};
};

export default getRouteAndMatch;
