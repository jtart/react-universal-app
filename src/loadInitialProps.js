const loadInitialProps = async (route, ctx) => {
  if (!route.getInitialProps) {
    return {};
  }

  try {
    return await route.getInitialProps(ctx);
  } catch (error) {
    throw error;
  }
};

export default loadInitialProps;
