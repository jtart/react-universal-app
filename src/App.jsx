import React, { Component } from 'react';
import { renderRoutes } from 'react-router-config';
import { withRouter } from 'react-router-dom';

import getRouteAndMatch from './getRouteAndMatch.js';
import loadInitialProps from './loadInitialProps.js';

export class App extends Component {
  constructor(props) {
    super(props);

    this.state = { data: this.props.initialData };
  }

  async componentDidUpdate({ location: prevLocation }) {
    if (this.props.location !== prevLocation) {
      const { route, match } = getRouteAndMatch(
        this.props.location.pathname,
        this.props.routes,
      );

      const data = await loadInitialProps(route, { match });

      this.setState({ data });
    }
  }

  render() {
    return renderRoutes(this.props.routes, { ...this.state.data });
  }
}

export default withRouter(App);
