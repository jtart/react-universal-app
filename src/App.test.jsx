import React from 'react';
import { App } from './App';
import * as reactRouterConfig from 'react-router-config';
import * as loadInitialProps from './loadInitialProps';
import * as getRouteAndMatch from './getRouteAndMatch.js';

describe('App', () => {
  let wrapper;
  let setStateSpy;
  let loadInitialPropsSpy;
  let getRouteAndMatchSpy;
  const initialData = { data: 'Some initial data' };

  reactRouterConfig.renderRoutes = jest
    .fn()
    .mockReturnValue(<h1>{initialData.data}</h1>);

  beforeAll(() => {
    wrapper = shallow(
      <App
        location={{ pathname: 'pathnameOne' }}
        routes={[]}
        initialData={initialData}
      />,
    );
    setStateSpy = jest.spyOn(wrapper.instance(), 'setState');
    loadInitialPropsSpy = spyOn(loadInitialProps, 'default');
    getRouteAndMatchSpy = spyOn(getRouteAndMatch, 'default');
  });

  it('should return rendered routes', () => {
    expect.assertions(5);
    expect(getRouteAndMatchSpy).not.toHaveBeenCalled();
    expect(loadInitialPropsSpy).not.toHaveBeenCalled();
    expect(reactRouterConfig.renderRoutes).toHaveBeenCalledTimes(1);
    expect(reactRouterConfig.renderRoutes).toHaveBeenCalledWith([], {
      data: initialData,
      error: null,
      loading: false,
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidUpdate', () => {
    describe('same location', () => {
      it('should not call set state with new data', () => {
        wrapper.setProps({ location: { pathname: 'pathnameOne' } });

        expect.assertions(3);
        expect(getRouteAndMatchSpy).not.toHaveBeenCalled();
        expect(loadInitialPropsSpy).not.toHaveBeenCalled();
        expect(setStateSpy).not.toHaveBeenCalled();
      });
    });

    describe('different location', () => {
      describe('no route or match', () => {
        it('should only call getRouteAndMatch', () => {
          const pathname = 'pathnameTwo';

          getRouteAndMatch.default = jest.fn().mockReturnValue({});

          wrapper.setProps({ location: { pathname } });

          expect.assertions(3);
          expect(getRouteAndMatch.default).toHaveBeenCalledWith(pathname, []);
          expect(loadInitialPropsSpy).not.toHaveBeenCalled();
          expect(setStateSpy).not.toHaveBeenCalled();
        });
      });

      describe('rejected loadInitialProps', () => {
        it('should console log the error', async () => {
          const error = 'Error!';
          spyOn(global.console, 'log');

          getRouteAndMatch.default = jest
            .fn()
            .mockReturnValue({ route: 'route', match: 'match' });
          loadInitialProps.default = jest
            .fn()
            .mockImplementation(() => Promise.reject(error));

          wrapper.setProps({ location: { pathname: 'pathnameThree' } });

          await loadInitialProps.default;

          expect.assertions(2);
          expect(global.console.log).toBeCalledWith(error);
          expect(setStateSpy).not.toBeCalled();
        });
      });

      describe('successful fetch of route, match, and initial props', () => {
        it('should call set state with new data', async () => {
          const pathname = 'pathnameFour';
          const route = 'some route';
          const match = 'a match';
          const data = 'Really cool data';

          getRouteAndMatch.default = jest
            .fn()
            .mockReturnValue({ route, match });
          loadInitialProps.default = jest
            .fn()
            .mockImplementation(async () => data);

          wrapper.setProps({ location: { pathname } });

          await loadInitialProps.default;

          expect.assertions(2);
          expect(loadInitialProps.default).toHaveBeenCalledWith(route, {
            match,
          });
          expect(setStateSpy).toBeCalledWith({ data });
        });
      });
    });
  });
});
