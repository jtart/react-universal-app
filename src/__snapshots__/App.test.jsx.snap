// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should return rendered routes 1`] = `
<h1>
  Some initial data
</h1>
`;
